### PR TITLE
PR update

### PR DESCRIPTION
**Pull request by vtols**
* Code refactoring
* New naming scheme
And more

### CODE REVIEW

File: src/collectd.conf.in [hunk 1]
```diff
@@ -1612,9 +1612,12 @@
...
 #<Plugin write_atsd>
-#     <Node "example">
-#         AtsdUrl "atsd_url"
-#     </Node>
+#  <Node "example">
+#    AtsdUrl "tcp://localhost:8081"
+#    Entity "example-entity"
+#    Prefix "collectd"
+#    ShortHostname false
+#  </Node>
```
Comment (rmakulov) [inline src/collectd.conf.in]: Please, add cache example.

File: src/utils_format_atsd.c [hunk 2]
```diff
@@ -16,22 +16,116 @@
...
+}
+
+static size_t strlcat(char *dst, const char *src, size_t siz) {
   char *d = dst;
   const char *s = src;
   size_t n = siz;
   size_t dlen;
 
-  /* Find the end of dst and adjust bytes left but don't go past end */
+  /* Find the end of dst and adjust bytes left but dobreak;n't go past end */
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: typo

File: src/utils_format_atsd.c [hunk 3]
```diff
@@ -51,75 +145,352 @@ size_t strlcat(char *dst, const char *src, size_t siz) {
...
+  if (format->ds->ds[format->index].type == DS_TYPE_GAUGE) {
+    BUFFER_ADD(GAUGE_FORMAT, format->vl->values[format->index].gauge);
+  } else if (format->rates != NULL) {
+    BUFFER_ADD("%.15g", format->rates[format->index]);
+  } else if (format->ds->ds[format->index].type == DS_TYPE_COUNTER) {
+    BUFFER_ADD("%llu", format->vl->values[format->index].counter);
+  } else if (format->ds->ds[format->index].type == DS_TYPE_DERIVE) {
+    BUFFER_ADD("%" PRIi64, format->vl->values[format->index].derive);
+  } else if (format->ds->ds[format->index].type == DS_TYPE_ABSOLUTE) {
+    BUFFER_ADD("%" PRIu64, format->vl->values[format->index].absolute);
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: Do we need to log error if unknown data source type was received?

File: src/utils_format_atsd.c [hunk 4]
```diff
@@ -51,75 +145,352 @@ size_t strlcat(char *dst, const char *src, size_t siz) {
...
+  return 0;
 }
 
 static int starts_with(const char *pre, const char *str) {
   size_t lenpre = strlen(pre), lenstr = strlen(str);
   return lenstr < lenpre ? 0 : strncmp(pre, str, lenpre) == 0;
 }
 
-void check_entity(char *ret, const int ret_len, const char *entity,
+int format_entity(char *ret, const int ret_len, const char *entity,
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: Please, make it a void function if no error handling.

Comment (vtols) [reply src/utils_format_atsd.c]: There is `strdup`, it can fail..

File: src/utils_format_atsd.c [hunk 5]
```diff
@@ -51,75 +145,352 @@ size_t strlcat(char *dst, const char *src, size_t siz) {
...
+                  true, invert_percent);
+    count++;
+    series_buffer++;
+  } else if (strcasecmp(format->vl->plugin, "df") == 0 &&
+             strcasecmp(format->vl->type, "percent_bytes") == 0 &&
+             strcasecmp(format->vl->type_instance, "free") == 0) {
+    format_series(series_buffer, format,
+                  NAME_PATTERN_PTR(PLUGIN, TYPE, STRING("used_reserved")), true,
+                  invert_percent);
+    count++;
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: If available get original instance from meta.

File: src/utils_format_atsd.c [hunk 6]
```diff
@@ -51,75 +145,352 @@ size_t strlcat(char *dst, const char *src, size_t siz) {
...
+
+  return ret;
+}
+
+size_t derive_series(series_t *series_buffer, format_info_t *format) {
+  char *key, *value, *strtok_ctx, *tmp;
+  _Bool preserve_original = true;
+  size_t count = 0;
+
+  if (format->rates != NULL && strcasecmp(format->vl->plugin, "cpu") == 0 &&
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: Add description

File: src/write_atsd.c [hunk 7]
```diff
@@ -17,33 +17,35 @@
  * Based on the write_graphite plugin.
  **/
 
-/* write_atsd plugin configuration example
+/* write_atsd plugin configuation example
```
Comment (rmakulov) [inline src/write_atsd.c]: typo

File: src/write_atsd.c [hunk 8]
```diff
@@ -17,33 +17,35 @@
  * Based on the write_graphite plugin.
```
Comment (rmakulov) [inline src/write_atsd.c]: Static functions

File: src/write_atsd.c [hunk 9]
```diff
@@ -346,44 +344,67 @@ static void wa_callback_free(void *data) {
...
-
-    sfree(atsd_stored_key->plugin);
-    sfree(atsd_stored_key->plugin_instance);
-    sfree(atsd_stored_key->type);
-    sfree(atsd_stored_key->type_instance);
-    sfree(atsd_stored_key);
-
-    sfree(atsd_stored_value->value);
-    sfree(atsd_stored_value);
+  if (cb->name != NULL)
```
Comment (rmakulov) [inline src/write_atsd.c]: Do we really need this checks?

File: src/utils_format_atsd.c [hunk 10]
```diff
@@ -51,75 +149,428 @@ size_t strlcat(char *dst, const char *src, size_t siz) {
...
+      if (ret != 0)
+        return -1;
+      preserve_original = false;
+    }
+
+    /* Fetch original unescaped disk name from meta */
+    if (format->vl->meta != NULL) {
+      char *disk_name;
+      ret = meta_data_get_string(format->vl->meta, "df:disk_name", &disk_name);
+      if (ret == 0) {
```
Comment (rmakulov) [inline src/utils_format_atsd.c]: df_disk_name or df:unescaped_plugin_instance?

File: src/write_atsd.c (hunk 9 above)
Comment (vtols) [reply src/write_atsd.c]: Yes, we need them, if initialization fials

File: src/utils_format_atsd.c (hunk 10 above)
Comment (vtols) [reply src/utils_format_atsd.c]: What do you prefer personally? Or maybe you have another variation

**Pull request merged**